pow/core/src/main/java/org/ethereum/beacon/pow/MinimalMerkle.java

```diff
@@ -0,0 +1,103 @@
+package org.ethereum.beacon.pow;
+
+import tech.pegasys.artemis.ethereum.core.Hash32;
+import tech.pegasys.artemis.util.bytes.Bytes32;
+import tech.pegasys.artemis.util.bytes.BytesValue;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static tech.pegasys.artemis.util.bytes.BytesValue.concat;
+
+/**
+ * Minimal merkle tree <a
+ * href="https://en.wikipedia.org/wiki/Merkle_tree">https://en.wikipedia.org/wiki/Merkle_tree</a>
+ * implementation
+ */
+public class MinimalMerkle {
```

Comment:
It looks like utility class for Merkle proof calculation. I remember your intention to swap it with an implementation that is used by SSZ hasher. It would be nice to have `MerkleTree` interface either.

Reply:
I've checked it and I don't think it's a good idea, as one in hasher holds the tree but this one provides only utility classes as there are a lot of slicing operations here. Plus proofs feature will be  only here. I could extract interface later, when incremental version is added and keep both,  but I don't think we could have one interface with `org.ethereum.beacon.ssz.visitor.MerkleTrie`